common/kvstore/ttl/ttl_wrapper.go

```diff
@@ -0,0 +1,303 @@
+// each expiry key points to is the regular key that is to be deleted when the expiry time is reached. In order to
+// efficiently delete expired keys, the expiry keys must be iterated over periodically to find and delete expired keys.
+type ttlStore struct {
+	store  kvstore.Store
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	logger logging.Logger
+}
+
+// TTLWrapper extends the given store with TTL capabilities. Periodically checks for expired keys and deletes them
+// with a period of gcPeriod. If gcPeriod is 0, no background goroutine is spawned to check for expired keys.
+//
+// Note: it is unsafe to access the wrapped store directly while the TTLStore is in use. The TTLStore uses special
+// key formatting, and direct access to the wrapped store may violate the TTLStore's invariants, resulting in
+// undefined behavior.
+func TTLWrapper(
+	ctx context.Context,
+	logger logging.Logger,
+	store kvstore.Store,
+	gcPeriod time.Duration) kvstore.TTLStore {
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	ttlStore := &ttlStore{
+		store:  store,
+		ctx:    ctx,
+		cancel: cancel,
+		logger: logger,
+	}
+	if gcPeriod > 0 {
+		ttlStore.expireKeysInBackground(gcPeriod)
+	}
+	return ttlStore
+}
+
+var keyPrefix = []byte("k")
+var expiryPrefix = []byte("e")
+var maxDeletionBatchSize = 1000
+
+// PutWithTTL adds a key-value pair to the store that expires after a specified time-to-live (TTL).
+// Key is eventually deleted after the TTL elapses.
+func (store *ttlStore) PutWithTTL(key []byte, value []byte, ttl time.Duration) error {
+	expiryTime := time.Now().Add(ttl)
+	return store.PutWithExpiration(key, value, expiryTime)
+}
+
+// PutBatchWithTTL adds multiple key-value pairs to the store that expire after a specified time-to-live (TTL).
+func (store *ttlStore) PutBatchWithTTL(keys [][]byte, values [][]byte, ttl time.Duration) error {
+	expiryTime := time.Now().Add(ttl)
+	return store.PutBatchWithExpiration(keys, values, expiryTime)
+}
+
+// buildExpiryKey creates an expiry key from the given expiry time.
+// The expiry key is composed of the following 3 components appended one after the other:
+// - a one byte "e" prefix
+// - the expiry time in hexadecimal format (8 bytes)
+// - and the base key.
+func buildExpiryKey(
+	baseKey []byte,
+	expiryTime time.Time) []byte {
+
+	expiryKeyLength := 1 /* prefix */ + 8 /* expiry timestamp */ + len(baseKey)
+	expiryKey := make([]byte, expiryKeyLength)
+
+	expiryKey[0] = 'e'
+	expiryUnixNano := expiryTime.UnixNano()
+	binary.BigEndian.PutUint64(expiryKey[1:], uint64(expiryUnixNano))
+
+	copy(expiryKey[9:], baseKey)
+
+	return expiryKey
+}
+
+// parseExpiryKey extracts the expiry time and base key from the given expiry key.
+func parseExpiryKey(expiryKey []byte) (baseKey []byte, expiryTime time.Time) {
+	expiryUnixNano := int64(0)
+	for i := 0; i < 8; i++ {
+		expiryUnixNano |= int64(expiryKey[1+i]) << (56 - i*8)
```

Comment:
The parsing needs to change corresponding to `PutUint64` as well.

Actually, it may be good to have test coverage for key building.

Reply:
One step ahead of you on this one. I already have this test. ;) The byte format is the same between the previous and new implementation, so this old line of code continued passing the test.

I've switched this implementation to instead use `binary.BigEndian.Uint64()` and verified that the unit test continues to pass.